plugins/node/opentelemetry-instrumentation-memcached/src/instrumentation.ts

```diff
@@ -0,0 +1,183 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as api from '@opentelemetry/api';
+import {
+  isWrapped,
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+} from '@opentelemetry/instrumentation';
+import type * as Memcached from 'memcached';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import * as utils from './utils';
+import { InstrumentationConfig } from './types';
+import { VERSION } from './version';
+
+export class Instrumentation extends InstrumentationBase<typeof Memcached> {
+  static readonly COMPONENT = 'memcached';
+  static readonly COMMON_ATTRIBUTES = {
+    [SemanticAttributes.DB_SYSTEM]: Instrumentation.COMPONENT,
+  };
+  static readonly DEFAULT_CONFIG: InstrumentationConfig = {
+    collectCommand: false,
+  };
+
+  constructor(config: InstrumentationConfig = Instrumentation.DEFAULT_CONFIG) {
+    super(
+      '@opentelemetry/instrumentation-memcached',
+      VERSION,
+      Object.assign({}, Instrumentation.DEFAULT_CONFIG, config)
+    );
+  }
+
+  setConfig(config: InstrumentationConfig = Instrumentation.DEFAULT_CONFIG) {
+    this._config = Object.assign({}, Instrumentation.DEFAULT_CONFIG, config);
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<typeof Memcached>(
+        'memcached',
+        ['>=2.2'],
+        (moduleExports, moduleVersion) => {
+          api.diag.debug(
+            `Patching ${Instrumentation.COMPONENT}@${moduleVersion}`
+          );
+          this.ensureWrapped(
+            moduleVersion,
+            moduleExports.prototype,
+            'command',
+            this.wrapCommand.bind(this, moduleVersion)
+          );
+          return moduleExports;
+        },
+        (moduleExports, moduleVersion) => {
+          api.diag.debug(
+            `Unpatching ${Instrumentation.COMPONENT}@${moduleVersion}`
+          );
+          if (moduleExports === undefined) return;
+          // `command` is documented API missing from the types
+          this._unwrap(moduleExports.prototype, 'command' as keyof Memcached);
+        }
+      ),
+    ];
+  }
+
+  wrapCommand(
+    moduleVersion: undefined | string,
+    original: (
+      queryCompiler: () => Memcached.CommandData,
+      server?: string
+    ) => any
+  ) {
+    const instrumentation = this;
+    return function (
+      this: Memcached,
+      queryCompiler: () => Memcached.CommandData,
+      server?: string
+    ) {
+      if (typeof queryCompiler !== 'function') {
+        return original.apply(this, arguments as any);
+      }
+      // The name will be overwritten later
+      const span = instrumentation.tracer.startSpan('<type> <key>', {
```

Comment:
I'd prefer `'unknown'` or `'unknown memcached command'` or something similar in case renaming somehow fails.

Reply:
Me too actually!